## Libraries

In [1]:
#!pip install autogluon

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 229.1/229.1 kB 4.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 kB 6.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 306.0/306.0 kB 10.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 416.7/416.7 kB 10.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 120.5/120.5 kB 14.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 25.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.3/12.3 MB 51.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.3/139.3 kB 14.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.4/64.4 kB 7.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.9/56.9 MB 17.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 64.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 619.9/619.9 MB 1.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━

In [1]:
import pandas as pd

from autogluon.tabular import TabularDataset, TabularPredictor

## Data preparation

Read in data from prepared file.

In [2]:
df = pd.read_csv('df_prepped.csv')
pd.set_option('display.max_columns', None)
df

Year Countries  Sand_1  Sand_2  Sand_3  Sand_4  Sand_5  Sand_6  Sand_7  \
0      2007    Angola      50      51      51      48      45      46      46   
1      2007    Angola      62      64      63      59      58      59      59   
2      2007    Angola      69      71      70      67      65      65      66   
3      2007    Angola      60      63      61      57      53      53      53   
4      2007    Angola      67      69      68      63      61      61      61   
...     ...       ...     ...     ...     ...     ...     ...     ...     ...   
32325  2016  Zimbabwe      73      75      74      69      66      66      66   
32326  2016  Zimbabwe      57      58      58      54      52      51      52   
32327  2016  Zimbabwe      55      57      56      54      52      51      51   
32328  2016  Zimbabwe      70      72      71      67      63      63      62   
32329  2016  Zimbabwe      62      64      63      60      59      60      60   

       Clay_1  Clay_2  Clay_3  Clay_4  Clay_5  Clay_6  Clay_7  OC_1  OC_2  \
0          37      35      36      39      42      42      42  0.52  0.23   
1          27      25      26      29      31      30      30  0.11  0.05   
2          19      16      18      21      24      24      23  0.09  0.06   
3          29      26      28      32      35      36      36  0.46  0.16   
4          22      19      21      25      28      28      29  0.15  0.09   
...       ...     ...     ...     ...     ...     ...     ...   ...   ...   
32325      20      18      20      24      27      27      27  0.10  0.08   
32326      28      26      26      31      33      34      33  0.14  0.12   
32327      30      28      30      32      34      35      35  0.53  0.41   
32328      17      14      15      20      24      24      24  0.14  0.09   
32329      24      22      24      27      29      28      28  0.08  0.06   

       OC_3  OC_4  OC_5  OC_6  OC_7  PAW_1  PAW_2  PAW_3  PAW_4  PAW_5  PAW_6  \
0      0.17  0.09  0.04  0.02  0.02   0.15   0.15   0.14   0.13   0.10   0.07   
1      0.07  0.04  0.02  0.02  0.01   0.11   0.10   0.10   0.09   0.07   0.07   
2      0.07  0.04  0.02  0.02  0.02   0.10   0.10   0.10   0.09   0.07   0.07   
3      0.14  0.08  0.05  0.04  0.03   0.12   0.13   0.12   0.12   0.11   0.10   
4      0.09  0.05  0.02  0.01  0.01   0.11   0.11   0.11   0.11   0.08   0.04   
...     ...   ...   ...   ...   ...    ...    ...    ...    ...    ...    ...   
32325  0.08  0.04  0.02  0.01  0.01   0.09   0.10   0.09   0.09   0.07   0.04   
32326  0.10  0.05  0.03  0.01  0.01   0.13   0.14   0.13   0.11   0.09   0.03   
32327  0.30  0.22  0.18  0.16  0.15   0.18   0.19   0.17   0.15   0.14   0.14   
32328  0.10  0.06  0.04  0.03  0.03   0.10   0.10   0.10   0.10   0.09   0.09   
32329  0.06  0.03  0.02  0.01  0.01   0.11   0.11   0.11   0.09   0.08   0.04   

       PAW_7  Y_maize_major           Farm  Sow_Maize_month_int  \
0       0.07       0.615357     104_Angola                    9   
1       0.03       0.257656      99_Angola                    9   
2       0.07       4.286831     108_Angola                    9   
3       0.09       0.700384     102_Angola                    9   
4       0.04       0.553450      43_Angola                    9   
...      ...            ...            ...                  ...   
32325   0.04       0.674080  3874_Zimbabwe                    5   
32326   0.03       0.707797  3875_Zimbabwe                    5   
32327   0.14       0.805844  3859_Zimbabwe                    5   
32328   0.08       0.595883  3871_Zimbabwe                    5   
32329   0.04       0.846773  3869_Zimbabwe                    5   

       Harvest_Maize_month_int  sow_to_harvest_months  maize_lag-1  \
0                            4                      7     0.554392   
1                            4                      7     0.117051   
2                            4                      7     3.093239   
3                            4                     

For faster demo testing, create subsampling. Default is whole dataset.

In [3]:
subsample_size = len(df)  # subsample subset of data for faster demo; full dataset (32K) fit takes appr 18 min, sample of 3K takes 1.5 min.
df = df.sample(n=subsample_size, random_state=0)

In [4]:
df.shape

(32330, 50)

Additionally, `Countries` and `Farm` attributes as categorical are dropped.

In [5]:
df = df.drop(['Countries','Farm'], axis=1)
df

Year  Sand_1  Sand_2  Sand_3  Sand_4  Sand_5  Sand_6  Sand_7  Clay_1  \
31671  2016      36      37      37      36      35      37      38      38   
5268   2008      63      65      65      63      60      60      59      24   
8384   2009      47      50      48      46      44      44      44      33   
230    2007      70      72      71      67      64      63      64      23   
17024  2011      55      57      55      50      47      47      47      28   
...     ...     ...     ...     ...     ...     ...     ...     ...     ...   
13123  2010      25      26      26      25      25      25      26      48   
19648  2012      36      36      36      34      33      33      33      37   
9845   2009      41      43      43      41      40      40      40      34   
10799  2009      53      56      55      52      47      47      47      26   
2732   2007      63      65      63      60      59      59      59      21   

       Clay_2  Clay_3  Clay_4  Clay_5  Clay_6  Clay_7  OC_1  OC_2  OC_3  OC_4  \
31671      37      39      38      40      39      38  0.36  0.27  0.17  0.09   
5268       22      23      25      28      28      28  0.51  0.21  0.18  0.10   
8384       30      32      36      39      40      38  0.34  0.21  0.16  0.09   
230        20      21      26      29      30      29  0.21  0.16  0.10  0.06   
17024      26      28      33      37      37      37  0.22  0.15  0.14  0.09   
...       ...     ...     ...     ...     ...     ...   ...   ...   ...   ...   
13123      47      47      50      52      53      52  0.56  0.36  0.29  0.21   
19648      36      38      41      44      45      46  0.79  0.38  0.30  0.22   
9845       32      32      33      37      37      36  0.75  0.52  0.33  0.23   
10799      23      25      28      34      35      33  0.32  0.16  0.13  0.07   
2732       20      21      24      25      25      25  0.20  0.16  0.16  0.09   

       OC_5  OC_6  OC_7  PAW_1  PAW_2  PAW_3  PAW_4  PAW_5  PAW_6  PAW_7  \
31671  0.06  0.04  0.04   0.14   0.14   0.13   0.13   0.11   0.10   0.10   
5268   0.09  0.07  0.06   0.12   0.12   0.12   0.11   0.11   0.11   0.11   
8384   0.04  0.03  0.03   0.14   0.15   0.14   0.13   0.11   0.09   0.09   
230    0.02  0.02  0.01   0.12   0.12   0.12   0.11   0.08   0.08   0.04   
17024  0.07  0.06  0.06   0.14   0.14   0.14   0.13   0.11   0.11   0.10   
...     ...   ...   ...    ...    ...    ...    ...    ...    ...    ...   
13123  0.17  0.15  0.14   0.17   0.17   0.16   0.15   0.13   0.13   0.13   
19648  0.17  0.15  0.14   0.17   0.19   0.17   0.16   0.15   0.14   0.14   
9845   0.20  0.16  0.17   0.20   0.20   0.17   0.16   0.14   0.14   0.14   
10799  0.04  0.03  0.03   0.14   0.15   0.14   0.13   0.12   0.11   0.11   
2732   0.07  0.05  0.05   0.13   0.14   0.13   0.12   0.11   0.10   0.10   

       Y_maize_major  Sow_Maize_month_int  Harvest_Maize_month_int  \
31671       1.404263                    3                        8   
5268        0.582666                    7                        1   
8384        0.903740                    4                        9   
230         0.488761                    9                        4   
17024       0.874741                   11                        4   
...              ...                  ...                      ...   
13123       2.589648                    3                       12   
19648       5.397777                    3                       12   
9845        3.343015                   11                        4   
10799       3.026406                    3                        8   
2732        0.472527                   10                        5   

       sow_to_harvest_months  maize_lag-1  pcp_mean_lag-1  tmax_mean_lag-1  \
31671                      5     1.331958       65.930227       307.842374   
5268                       6     0.613687      121.964673       302.328427   
8384                       5     0.904677      119.691173       303.917189   
230                

Create train and test sets.

In [6]:
# Separate a test set, the year 2016
df_test = df[df.Year == 2016]
df_train = df[df.Year < 2016]

In [7]:
df_test.shape, df_train.shape

((2930, 48), (29400, 48))

## AutoGluon: fit, predict, feature importance

Reference: https://auto.gluon.ai/stable/tutorials/tabular/tabular-essentials.html

Dependent variable for this regression task is `Y_maize_major`.

In [8]:
label = 'Y_maize_major'
df_train[label].describe()

count    29400.000000
mean         1.716275
std          1.548682
min          0.000000
25%          0.790347
50%          1.227576
75%          2.169080
max         14.542289
Name: Y_maize_major, dtype: float64

### Fit and predict on RMSE

Initialize and fit AutoGluon TabularPredictor. As two metrics - RMSE and MAE - are used in this project, these need to be fitted separately. Default metric for AutoGluon is RMSE.

In [9]:
# takes about 18 minutes
metric = 'root_mean_squared_error' # rmse is default metric

predictor_rmse = TabularPredictor(label=label, eval_metric=metric).fit(df_train)

No path specified. Models will be saved in: "AutogluonModels/ag-20231215_153237"
No presets specified! To achieve strong results with AutoGluon, it is recommended to use the available presets.
	Recommended Presets (For more details refer to https://auto.gluon.ai/stable/tutorials/tabular/tabular-essentials.html#presets):
	presets='best_quality'   : Maximize accuracy. Default time_limit=3600.
	presets='high_quality'   : Strong accuracy with fast inference speed. Default time_limit=3600.
	presets='good_quality'   : Good accuracy with very fast inference speed. Default time_limit=3600.
	presets='medium_quality' : Fast training time, ideal for initial prototyping.
Beginning AutoGluon training ...
AutoGluon will save models to "AutogluonModels/ag-20231215_153237"
=================== System Info ===================
AutoGluon Version:  1.0.0
Python Version:     3.10.12
Operating System:   Linux
Platform Machine:   x86_64
Platform Version:   #1 SMP PREEMPT_DYNAMIC Sat Nov 18 15:31:17 UTC 2023
C

[1000]	valid_set's rmse: 0.255532
[2000]	valid_set's rmse: 0.248871
[3000]	valid_set's rmse: 0.245956
[4000]	valid_set's rmse: 0.244403
[5000]	valid_set's rmse: 0.243261
[6000]	valid_set's rmse: 0.242666
[7000]	valid_set's rmse: 0.242099
[8000]	valid_set's rmse: 0.241796
[9000]	valid_set's rmse: 0.241691
[10000]	valid_set's rmse: 0.24163


	-0.2416	 = Validation score   (-root_mean_squared_error)
	54.1s	 = Training   runtime
	4.56s	 = Validation runtime
Fitting model: LightGBM ...


[1000]	valid_set's rmse: 0.243889
[2000]	valid_set's rmse: 0.240797
[3000]	valid_set's rmse: 0.239857
[4000]	valid_set's rmse: 0.23966
[5000]	valid_set's rmse: 0.239605
[6000]	valid_set's rmse: 0.239482
[7000]	valid_set's rmse: 0.239432
[8000]	valid_set's rmse: 0.239414


	-0.2393	 = Validation score   (-root_mean_squared_error)
	56.0s	 = Training   runtime
	1.82s	 = Validation runtime
Fitting model: RandomForestMSE ...
	-0.2781	 = Validation score   (-root_mean_squared_error)
	335.1s	 = Training   runtime
	0.41s	 = Validation runtime
Fitting model: CatBoost ...
	-0.2343	 = Validation score   (-root_mean_squared_error)
	257.58s	 = Training   runtime
	0.03s	 = Validation runtime
Fitting model: ExtraTreesMSE ...
	-0.2719	 = Validation score   (-root_mean_squared_error)
	89.29s	 = Training   runtime
	0.36s	 = Validation runtime
Fitting model: NeuralNetFastAI ...
	-0.266	 = Validation score   (-root_mean_squared_error)
	40.51s	 = Training   runtime
	0.06s	 = Validation runtime
Fitting model: XGBoost ...
	-0.2435	 = Validation score   (-root_mean_squared_error)
	83.2s	 = Training   runtime
	2.9s	 = Validation runtime
Fitting model: NeuralNetTorch ...
	-0.297	 = Validation score   (-root_mean_squared_error)
	67.72s	 = Training   runtime
	0.04s	 = Validation r

[1000]	valid_set's rmse: 0.24768
[2000]	valid_set's rmse: 0.247121
[3000]	valid_set's rmse: 0.246846
[4000]	valid_set's rmse: 0.246707
[5000]	valid_set's rmse: 0.246616
[6000]	valid_set's rmse: 0.246586
[7000]	valid_set's rmse: 0.246563
[8000]	valid_set's rmse: 0.246555
[9000]	valid_set's rmse: 0.246549
[10000]	valid_set's rmse: 0.246546


	-0.2465	 = Validation score   (-root_mean_squared_error)
	142.58s	 = Training   runtime
	7.91s	 = Validation runtime
Fitting model: WeightedEnsemble_L2 ...
	Ensemble Weights: {'CatBoost': 0.414, 'LightGBMXT': 0.287, 'XGBoost': 0.138, 'RandomForestMSE': 0.08, 'LightGBM': 0.069, 'NeuralNetFastAI': 0.011}
	-0.229	 = Validation score   (-root_mean_squared_error)
	0.41s	 = Training   runtime
	0.01s	 = Validation runtime
AutoGluon training complete, total runtime = 1178.34s ... Best model: "WeightedEnsemble_L2"
TabularPredictor saved. To load, use: predictor = TabularPredictor.load("AutogluonModels/ag-20231215_153237")


Now our trained model is used to make predictions on the new data:

In [10]:
y_pred_rmse = predictor_rmse.predict(df_test.drop(columns=[label]))
y_pred_rmse.head()

31671    1.430020
31429    3.766737
31523    2.538168
31574    3.715868
30025    0.740948
Name: Y_maize_major, dtype: float32

Next, we can evaluate the predictor on the (labeled) test data:

In [11]:
predictor_rmse.evaluate(df_test, silent=True)

{'root_mean_squared_error': -0.33155242006856295,
 'mean_squared_error': -0.10992700725332083,
 'mean_absolute_error': -0.21335905398448385,
 'r2': 0.958838550687088,
 'pearsonr': 0.9809962523591725,
 'median_absolute_error': -0.1289768517853241}

For evaluation metrics where higher values are worse (like RMSE), AutoGluon will flip their sign and print them as negative values during training (as it internally assumes higher values are better).

Each model is evaluated individually, in this case based on RMSE.

In [12]:
predictor_rmse.leaderboard(df_test)

model  score_test  score_val              eval_metric  \
0         ExtraTreesMSE   -0.306964  -0.271904  root_mean_squared_error   
1   WeightedEnsemble_L2   -0.331552  -0.229048  root_mean_squared_error   
2        NeuralNetTorch   -0.338696  -0.297004  root_mean_squared_error   
3       NeuralNetFastAI   -0.341045  -0.266038  root_mean_squared_error   
4            LightGBMXT   -0.347832  -0.241606  root_mean_squared_error   
5       RandomForestMSE   -0.352906  -0.278077  root_mean_squared_error   
6               XGBoost   -0.364392  -0.243472  root_mean_squared_error   
7              LightGBM   -0.376283  -0.239337  root_mean_squared_error   
8         LightGBMLarge   -0.376523  -0.246546  root_mean_squared_error   
9              CatBoost   -0.377605  -0.234307  root_mean_squared_error   
10       KNeighborsDist   -0.825696  -0.726367  root_mean_squared_error   
11       KNeighborsUnif   -0.839621  -0.749136  root_mean_squared_error   

    pred_time_test  pred_time_val    fit_time  pred_time_test_marginal  \
0         0.718004       0.362677   89.294094                 0.718004   
1        12.888773       9.788903  826.899203                 0.026028   
2         0.046891       0.035445   67.720575                 0.046891   
3         0.130257       0.058580   40.509411                 0.130257   
4         5.027999       4.559393   54.103845                 5.027999   
5         1.081275       0.412323  335.099097                 1.081275   
6         3.061738       2.900748   83.199527                 3.061738   
7         3.512791       1.818392   55.999351                 3.512791   
8        11.128875       7.908750  142.575995                11.128875   
9         0.048686       0.031953  257.577009                 0.048686   
10        0.624105       0.527812    0.045444                 0.624105   
11        0.628608       0.607425    3.161382                 0.628608   

    pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  \
0                 0.362677          89.294094            1       True   
1                 0.007514           0.410964            2       True   
2                 0.035445          67.720575            1       True   
3                 0.058580          40.509411            1       True   
4                 4.559393          54.103845            1       True   
5                 0.412323         335.099097            1       True   
6                 2.900748          83.199527            1       True   
7                 1.818392          55.999351            1       True   
8                 7.908750         142.575995            1       True   
9                 0.031953         257.577009            1       True   
10                0.527812           0.045444            1       True   
11                0.607425           3.161382            1       True   

    fit_order  
0           7  
1          12  
2          10  
3           8  
4           3  
5           5  
6           9  
7           4  
8          11  
9           6  
10          2  
11          1

Best model is ExtraTreesMSE with RMSE on test set of 0.313.

For the whole model, RMSE on test data is 0.3801 and MAE 0.239.

### Fit and predict on MAE

Same procedure is now followed by using mean absolute error as metric. First fit on train data:

In [13]:
metric = 'mean_absolute_error'
predictor_mae = TabularPredictor(label=label, eval_metric=metric).fit(df_train)

No path specified. Models will be saved in: "AutogluonModels/ag-20231215_155310"
No presets specified! To achieve strong results with AutoGluon, it is recommended to use the available presets.
	Recommended Presets (For more details refer to https://auto.gluon.ai/stable/tutorials/tabular/tabular-essentials.html#presets):
	presets='best_quality'   : Maximize accuracy. Default time_limit=3600.
	presets='high_quality'   : Strong accuracy with fast inference speed. Default time_limit=3600.
	presets='good_quality'   : Good accuracy with very fast inference speed. Default time_limit=3600.
	presets='medium_quality' : Fast training time, ideal for initial prototyping.
Beginning AutoGluon training ...
AutoGluon will save models to "AutogluonModels/ag-20231215_155310"
=================== System Info ===================
AutoGluon Version:  1.0.0
Python Version:     3.10.12
Operating System:   Linux
Platform Machine:   x86_64
Platform Version:   #1 SMP PREEMPT_DYNAMIC Sat Nov 18 15:31:17 UTC 2023
C

[1000]	valid_set's l1: 0.143418
[2000]	valid_set's l1: 0.136363
[3000]	valid_set's l1: 0.133073
[4000]	valid_set's l1: 0.131139
[5000]	valid_set's l1: 0.12969
[6000]	valid_set's l1: 0.128753
[7000]	valid_set's l1: 0.128207
[8000]	valid_set's l1: 0.127636
[9000]	valid_set's l1: 0.127183
[10000]	valid_set's l1: 0.126907


	-0.1269	 = Validation score   (-mean_absolute_error)
	53.28s	 = Training   runtime
	4.75s	 = Validation runtime
Fitting model: LightGBM ...


[1000]	valid_set's l1: 0.131505
[2000]	valid_set's l1: 0.127306
[3000]	valid_set's l1: 0.125699
[4000]	valid_set's l1: 0.125165
[5000]	valid_set's l1: 0.124984
[6000]	valid_set's l1: 0.124868
[7000]	valid_set's l1: 0.124784
[8000]	valid_set's l1: 0.124692
[9000]	valid_set's l1: 0.124607
[10000]	valid_set's l1: 0.124568


	-0.1246	 = Validation score   (-mean_absolute_error)
	62.24s	 = Training   runtime
	3.2s	 = Validation runtime
Fitting model: RandomForestMSE ...
	-0.1492	 = Validation score   (-mean_absolute_error)
	328.91s	 = Training   runtime
	0.38s	 = Validation runtime
Fitting model: CatBoost ...
	-0.1224	 = Validation score   (-mean_absolute_error)
	343.29s	 = Training   runtime
	0.03s	 = Validation runtime
Fitting model: ExtraTreesMSE ...
	-0.1466	 = Validation score   (-mean_absolute_error)
	84.65s	 = Training   runtime
	0.35s	 = Validation runtime
Fitting model: NeuralNetFastAI ...
	-0.1534	 = Validation score   (-mean_absolute_error)
	40.06s	 = Training   runtime
	0.05s	 = Validation runtime
Fitting model: XGBoost ...
	-0.128	 = Validation score   (-mean_absolute_error)
	83.08s	 = Training   runtime
	2.62s	 = Validation runtime
Fitting model: NeuralNetTorch ...
	-0.1678	 = Validation score   (-mean_absolute_error)
	56.2s	 = Training   runtime
	0.04s	 = Validation runtime
Fitting model: Lig

[1000]	valid_set's l1: 0.126279
[2000]	valid_set's l1: 0.124518
[3000]	valid_set's l1: 0.124064
[4000]	valid_set's l1: 0.123882
[5000]	valid_set's l1: 0.123793
[6000]	valid_set's l1: 0.12376
[7000]	valid_set's l1: 0.123725
[8000]	valid_set's l1: 0.123717
[9000]	valid_set's l1: 0.123712
[10000]	valid_set's l1: 0.123709


	-0.1237	 = Validation score   (-mean_absolute_error)
	136.78s	 = Training   runtime
	8.74s	 = Validation runtime
Fitting model: WeightedEnsemble_L2 ...
	Ensemble Weights: {'LightGBMLarge': 0.308, 'CatBoost': 0.275, 'LightGBMXT': 0.253, 'LightGBM': 0.121, 'ExtraTreesMSE': 0.022, 'NeuralNetFastAI': 0.022}
	-0.1172	 = Validation score   (-mean_absolute_error)
	0.79s	 = Training   runtime
	0.0s	 = Validation runtime
AutoGluon training complete, total runtime = 1238.72s ... Best model: "WeightedEnsemble_L2"
TabularPredictor saved. To load, use: predictor = TabularPredictor.load("AutogluonModels/ag-20231215_155310")


Then make predictions:

In [14]:
y_pred_mae = predictor_mae.predict(df_test.drop(columns=[label]))
y_pred_mae.head()

31671    1.406964
31429    3.841526
31523    2.580132
31574    3.657898
30025    0.745997
Name: Y_maize_major, dtype: float32

Then evaluation on labeled test data:

In [15]:
predictor_mae.evaluate(df_test, silent=True)

{'mean_absolute_error': -0.21689106127634075,
 'root_mean_squared_error': -0.3385265964471238,
 'mean_squared_error': -0.11460025650207382,
 'r2': 0.9570886830532328,
 'pearsonr': 0.9805619952309511,
 'median_absolute_error': -0.13077324649505623}

And then model ranking based on MAE:

In [16]:
predictor_mae.leaderboard(df_test)

model  score_test  score_val          eval_metric  \
0         ExtraTreesMSE   -0.190160  -0.146551  mean_absolute_error   
1       RandomForestMSE   -0.208269  -0.149217  mean_absolute_error   
2        NeuralNetTorch   -0.213067  -0.167844  mean_absolute_error   
3   WeightedEnsemble_L2   -0.216891  -0.117197  mean_absolute_error   
4            LightGBMXT   -0.222291  -0.126903  mean_absolute_error   
5       NeuralNetFastAI   -0.225310  -0.153432  mean_absolute_error   
6               XGBoost   -0.230537  -0.128046  mean_absolute_error   
7         LightGBMLarge   -0.233662  -0.123709  mean_absolute_error   
8              LightGBM   -0.238951  -0.124552  mean_absolute_error   
9              CatBoost   -0.240287  -0.122411  mean_absolute_error   
10       KNeighborsDist   -0.527741  -0.465524  mean_absolute_error   
11       KNeighborsUnif   -0.539566  -0.479410  mean_absolute_error   

    pred_time_test  pred_time_val    fit_time  pred_time_test_marginal  \
0         0.719697       0.353073   84.649351                 0.719697   
1         1.218788       0.383575  328.911488                 1.218788   
2         0.045728       0.038885   56.204100                 0.045728   
3        22.289909      17.119061  721.094695                 0.031968   
4         6.615834       4.747211   53.276853                 6.615834   
5         0.148433       0.051335   40.060186                 0.148433   
6         4.398964       2.620618   83.078391                 4.398964   
7        10.859820       8.742219  136.781504                10.859820   
8         3.867173       3.195118   62.244542                 3.867173   
9         0.046985       0.029148  343.294945                 0.046985   
10        0.635215       0.542662    0.047155                 0.635215   
11        0.621442       0.542677    0.042695                 0.621442   

    pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  \
0                 0.353073          84.649351            1       True   
1                 0.383575         328.911488            1       True   
2                 0.038885          56.204100            1       True   
3                 0.000958           0.787314            2       True   
4                 4.747211          53.276853            1       True   
5                 0.051335          40.060186            1       True   
6                 2.620618          83.078391            1       True   
7                 8.742219         136.781504            1       True   
8                 3.195118          62.244542            1       True   
9                 0.029148         343.294945            1       True   
10                0.542662           0.047155            1       True   
11                0.542677           0.042695            1       True   

    fit_order  
0           7  
1           5  
2          10  
3          12  
4           3  
5           8  
6           9  
7          11  
8           4  
9           6  
10          2  
11          1

Best model based on MAE is also ExtraTreesMSE with MAE on test data of 0.197.

### Feature importance

To better understand our trained predictor, we can estimate the overall importance of each feature via `feature_importance()`. First, model fitted based on RMSE:

In [17]:
predictor_rmse.feature_importance(df_test) #14 minutes

Computing feature importance via permutation shuffling for 47 features using 2930 rows with 5 shuffle sets...
	3441.87s	= Expected runtime (688.37s per shuffle set)
	2879.4s	= Actual runtime (Completed 5 of 5 shuffle sets)


importance    stddev       p_value  n  p99_high  \
maize_lag-1                0.888633  0.011201  3.028354e-09  5  0.911696   
maize_lag-3                0.344398  0.005511  7.867488e-09  5  0.355746   
maize_lag-2                0.333098  0.008041  4.072353e-08  5  0.349655   
tmax_mean_lag-2            0.016696  0.000789  5.970548e-07  5  0.018321   
tmax_mean_lag-3            0.013333  0.000421  1.193397e-07  5  0.014200   
Sow_Maize_month_int        0.003812  0.001451  2.098327e-03  5  0.006800   
tmin_mean_lag-3            0.003766  0.001220  1.156602e-03  5  0.006278   
spi_mean_lag-3             0.002339  0.001929  2.672328e-02  5  0.006310   
pcp_mean_lag-2             0.001743  0.000876  5.628223e-03  5  0.003547   
spi_mean_lag-2             0.001409  0.002001  9.519856e-02  5  0.005529   
OC_6                       0.001357  0.000545  2.549662e-03  5  0.002479   
OC_5                       0.001190  0.000365  9.436964e-04  5  0.001943   
PAW_6                      0.000964  0.000314  1.175202e-03  5  0.001610   
Clay_2                     0.000394  0.000352  3.327805e-02  5  0.001120   
spi_mean_lag-1             0.000252  0.001490  3.620511e-01  5  0.003321   
tmin_mean_lag-2            0.000196  0.000729  2.898240e-01  5  0.001697   
Year                       0.000000  0.000000  5.000000e-01  5  0.000000   
Clay_6                    -0.000038  0.000180  6.679644e-01  5  0.000334   
OC_3                      -0.000158  0.000154  9.582218e-01  5  0.000159   
pcp_mean_lag-3            -0.000397  0.000854  8.213374e-01  5  0.001362   
Clay_3                    -0.000471  0.000202  9.967652e-01  5 -0.000055   
Clay_4                    -0.000538  0.000247  9.958754e-01  5 -0.000029   
PAW_2                     -0.000618  0.000129  9.997847e-01  5 -0.000353   
tmin_mean_lag-1           -0.000708  0.001449  8.321639e-01  5  0.002275   
OC_7                      -0.000775  0.000158  9.998016e-01  5 -0.000449   
PAW_1                     -0.000788  0.000368  9.956283e-01  5 -0.000030   
PAW_7                     -0.000839  0.000135  9.999223e-01  5 -0.000561   
Clay_5                    -0.000895  0.000390  9.966006e-01  5 -0.000093   
OC_4                      -0.000981  0.000153  9.999319e-01  5 -0.000667   
OC_2                      -0.001341  0.000515  9.978416e-01  5 -0.000282   
PAW_5                     -0.001404  0.000358  9.995338e-01  5 -0.000667   
Sand_7                    -0.001467  0.000342  9.996704e-01  5 -0.000763   
Sand_6                    -0.001832  0.000220  9.999756e-01  5 -0.001379   
Sand_5                    -0.001866  0.000269  9.999498e-01  5 -0.001313   
tmax_mean_lag-1           -0.001927  0.001735  9.660191e-01  5  0.001645   
PAW_4                     -0.001935  0.000284  9.999461e-01  5 -0.001351   
Sand_2                    -0.002095  0.000236  9.999811e-01  5 -0.001610   
PAW_3                     -0.002108  0.000237  9.999812e-01  5 -0.001620   
pcp_mean_lag-1            -0.002314  0.000960  9.971362e-01  5 -0.000338   
Sand_4                    -0.002382  0.000504  9.997738e-01  5 -0.001345   
OC_1                      -0.002545  0.001033  9.973547e-01  5 -0.000419   
Clay_7                    -0.002718  0.000509  9.998589e-01  5 -0.001669   
Sand_3                    -0.003156  0.000243  9.999958e-01  5 -0.002656   
Clay_1                    -0.003180  0.000296  9.999911e-01  5 -0.002571   
Sand_1                    -0.003811  0.000943  9.995854e-01  5 -0.001870   
Harvest_Maize_month_int   -0.012507  0.001424  9.999802e-01  5 -0.009574   
sow_to_harvest_months     -0.021702  0.001370  9.999981e-01  5 -0.018880   

                          p99_low  
maize_lag-1              0.865570  
maize_lag-3              0.333050  
maize_lag-2              0.316541  
tmax_mean_lag-2          0.015071  
tmax_mean_lag-3          0.012466  
Sow_Maize_month_int      0.000824  
tmin_mean_lag-3          0.001253  
spi_mean_lag-3          -0.001632  
pcp_mean_lag-2          -0.000061  
spi_mean_lag-2  

Then, model fitted based on MAE:

In [18]:
predictor_mae.feature_importance(df_test)

Computing feature importance via permutation shuffling for 47 features using 2930 rows with 5 shuffle sets...
	5426.96s	= Expected runtime (1085.39s per shuffle set)
	4654.54s	= Actual runtime (Completed 5 of 5 shuffle sets)


importance    stddev       p_value  n  p99_high  \
maize_lag-1                0.584304  0.009757  9.325911e-09  5  0.604393   
maize_lag-2                0.186154  0.008100  4.290028e-07  5  0.202831   
maize_lag-3                0.134500  0.004810  1.959654e-07  5  0.144404   
tmax_mean_lag-2            0.007641  0.001101  5.035184e-05  5  0.009908   
tmax_mean_lag-3            0.006715  0.000718  1.541816e-05  5  0.008193   
tmin_mean_lag-3            0.002656  0.000749  6.853681e-04  5  0.004198   
spi_mean_lag-2             0.001678  0.001001  9.983107e-03  5  0.003738   
pcp_mean_lag-2             0.001062  0.000527  5.372457e-03  5  0.002146   
tmin_mean_lag-2            0.000851  0.000482  8.387555e-03  5  0.001843   
OC_5                       0.000786  0.000127  7.829227e-05  5  0.001047   
Clay_2                     0.000469  0.000437  3.712989e-02  5  0.001369   
PAW_6                      0.000391  0.000133  1.397829e-03  5  0.000666   
OC_6                       0.000077  0.000186  2.040214e-01  5  0.000460   
Year                       0.000000  0.000000  5.000000e-01  5  0.000000   
spi_mean_lag-1            -0.000038  0.000660  5.486270e-01  5  0.001321   
OC_7                      -0.000196  0.000165  9.716958e-01  5  0.000144   
spi_mean_lag-3            -0.000237  0.001311  6.463881e-01  5  0.002463   
OC_3                      -0.000239  0.000068  9.993013e-01  5 -0.000100   
PAW_1                     -0.000309  0.000169  9.925117e-01  5  0.000039   
pcp_mean_lag-3            -0.000311  0.000609  8.418772e-01  5  0.000941   
Sand_2                    -0.000337  0.000211  9.883910e-01  5  0.000097   
pcp_mean_lag-1            -0.000391  0.000601  8.904211e-01  5  0.000846   
OC_2                      -0.000429  0.000215  9.944819e-01  5  0.000012   
PAW_7                     -0.000431  0.000091  9.997782e-01  5 -0.000244   
PAW_2                     -0.000446  0.000076  9.999029e-01  5 -0.000290   
PAW_5                     -0.000459  0.000152  9.987645e-01  5 -0.000147   
Clay_5                    -0.000500  0.000129  9.995102e-01  5 -0.000234   
OC_4                      -0.000521  0.000181  9.985032e-01  5 -0.000148   
Clay_6                    -0.000535  0.000145  9.994082e-01  5 -0.000236   
Clay_4                    -0.000610  0.000115  9.998573e-01  5 -0.000374   
tmax_mean_lag-1           -0.000707  0.001197  8.712500e-01  5  0.001759   
PAW_4                     -0.000721  0.000169  9.996606e-01  5 -0.000372   
Clay_3                    -0.000842  0.000111  9.999642e-01  5 -0.000613   
Sand_5                    -0.000934  0.000142  9.999381e-01  5 -0.000642   
Sand_7                    -0.001114  0.000097  9.999931e-01  5 -0.000914   
Sand_3                    -0.001154  0.000120  9.999861e-01  5 -0.000907   
PAW_3                     -0.001406  0.000151  9.999843e-01  5 -0.001095   
Sand_6                    -0.001505  0.000101  9.999976e-01  5 -0.001296   
Sand_4                    -0.001545  0.000130  9.999941e-01  5 -0.001278   
OC_1                      -0.001725  0.000581  9.986680e-01  5 -0.000530   
Clay_7                    -0.001753  0.000210  9.999758e-01  5 -0.001321   
tmin_mean_lag-1           -0.002013  0.001095  9.926456e-01  5  0.000241   
Clay_1                    -0.002234  0.000248  9.999819e-01  5 -0.001722   
Sand_1                    -0.002332  0.000271  9.999785e-01  5 -0.001774   
Sow_Maize_month_int       -0.002795  0.000708  9.995453e-01  5 -0.001337   
Harvest_Maize_month_int   -0.008384  0.000624  9.999963e-01  5 -0.007099   
sow_to_harvest_months     -0.015904  0.000522  9.999999e-01  5 -0.014830   

                          p99_low  
maize_lag-1              0.564214  
maize_lag-2              0.169477  
maize_lag-3              0.124596  
tmax_mean_lag-2          0.005373  
tmax_mean_lag-3          0.005238  
tmin_mean_lag-3          0.001113  
spi_mean_lag-2          -0.000383  
pcp_mean_lag-2          -0.000022  
tmin_mean_lag-2         -0.000140  
OC_5            